File: lib/app/core/modules/material/controllers/material_controller.dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import '../../../../data/services/api_service.dart';
import '../../../../data/services/auth_service.dart';
import '../../../../core/constants/app_constants.dart';
import '../../../../data/models/material_model.dart';
import '../../../../data/models/project_model.dart';

class MaterialController extends GetxController {
  final apiService = Get.find<ApiService>();
  final authService = Get.find<AuthService>();

  final isLoading = false.obs;
  final materials = <MaterialModel>[].obs;
  final selectedMaterial = Rx<MaterialModel?>(null);

  // Form Controllers
  final materialNameController = TextEditingController();
  final quantityController = TextEditingController();
  final unitController = TextEditingController();
  final descriptionController = TextEditingController();
  final selectedProject = Rx<ProjectModel?>(null);

  @override
  void onClose() {
    materialNameController.dispose();
    quantityController.dispose();
    unitController.dispose();
    descriptionController.dispose();
    super.onClose();
  }

  // Fetch materials by project dengan filter role
  Future<void> fetchMaterialsByProject(String projectId) async {
    try {
      isLoading.value = true;

      final response = await apiService.get(
        '${AppConstants.materialsEndpoint}/project/$projectId',
      );

      if (response['status'] == 'success') {
        final allMaterials = (response['data'] as List)
            .map((json) => MaterialModel.fromJson(json))
            .toList();

        // Filter berdasarkan role jika perlu
        materials.value = _filterMaterialsByRole(allMaterials);
      }
    } catch (e) {
      _showError(e.toString());
    } finally {
      isLoading.value = false;
    }
  }

  List<MaterialModel> _filterMaterialsByRole(List<MaterialModel> allMaterials) {
    final user = authService.currentUser.value;
    if (user == null) return [];

    // Admin & Kepala Proyek: Lihat semua material
    if (authService.isAdmin || authService.isKepalaProyek) {
      return allMaterials;
    }

    // Mandor: Lihat material yang dia request
    if (authService.isMandor) {
      return allMaterials.where((m) => m.mandorId == user.id).toList();
    }

    // Pengguna: Lihat semua material dari project mereka (read-only)
    if (authService.isPengguna) {
      return allMaterials;
    }

    return [];
  }

  // Get material detail
  Future<void> getMaterialDetail(String id) async {
    try {
      isLoading.value = true;

      final response = await apiService.get(
        '${AppConstants.materialsEndpoint}/$id',
      );

      if (response['status'] == 'success') {
        selectedMaterial.value = MaterialModel.fromJson(response['data']);
      }
    } catch (e) {
      _showError(e.toString());
    } finally {
      isLoading.value = false;
    }
  }

  // Create material request (Mandor only)
  Future<void> createMaterialRequest() async {
    if (!authService.isMandor) {
      _showError('Hanya mandor yang dapat request material');
      return;
    }

    if (!_validateForm()) return;

    try {
      isLoading.value = true;

      final response = await apiService.post(
        AppConstants.materialsEndpoint,
        {
          'project_id': selectedProject.value!.id,
          'material_name': materialNameController.text.trim(),
          'quantity': double.parse(quantityController.text),
          'unit': unitController.text.trim(),
          'description': descriptionController.text.trim(),
        },
      );

      if (response['status'] == 'success') {
        _showSuccess('Request material berhasil dibuat');

        if (selectedProject.value != null) {
          await fetchMaterialsByProject(selectedProject.value!.id);
        }

        Get.back();
        _clearForm();
      }
    } catch (e) {
      _showError(e.toString());
    } finally {
      isLoading.value = false;
    }
  }

  // Approve material (Kepala Proyek / Admin ONLY)
  Future<void> approveMaterial(String id) async {
    if (!authService.isKepalaProyek && !authService.isAdmin) {
      _showError(
          'Hanya kepala proyek atau admin yang dapat menyetujui material');
      return;
    }

    try {
      final confirm = await Get.dialog<bool>(
        AlertDialog(
          title: const Text('Konfirmasi'),
          content: const Text('Setujui request material ini?'),
          actions: [
            TextButton(
              onPressed: () => Get.back(result: false),
              child: const Text('Batal'),
            ),
            ElevatedButton(
              onPressed: () => Get.back(result: true),
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.green,
              ),
              child: const Text('Setujui'),
            ),
          ],
        ),
      );

      if (confirm == true) {
        isLoading.value = true;

        final response = await apiService.put(
          '${AppConstants.materialsEndpoint}/$id/approve',
          {},
        );

        if (response['status'] == 'success') {
          _showSuccess('Material berhasil disetujui');

          await getMaterialDetail(id);

          // Refresh list if project selected
          if (selectedProject.value != null) {
            await fetchMaterialsByProject(selectedProject.value!.id);
          }
        }
      }
    } catch (e) {
      _showError(e.toString());
    } finally {
      isLoading.value = false;
    }
  }

  // Reject material (Kepala Proyek / Admin ONLY)
  Future<void> rejectMaterial(String id) async {
    if (!authService.isKepalaProyek && !authService.isAdmin) {
      _showError('Hanya kepala proyek atau admin yang dapat menolak material');
      return;
    }

    try {
      final confirm = await Get.dialog<bool>(
        AlertDialog(
          title: const Text('Konfirmasi'),
          content: const Text('Tolak request material ini?'),
          actions: [
            TextButton(
              onPressed: () => Get.back(result: false),
              child: const Text('Batal'),
            ),
            ElevatedButton(
              onPressed: () => Get.back(result: true),
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.red,
              ),
              child: const Text('Tolak'),
            ),
          ],
        ),
      );

      if (confirm == true) {
        isLoading.value = true;

        final response = await apiService.put(
          '${AppConstants.materialsEndpoint}/$id/reject',
          {},
        );

        if (response['status'] == 'success') {
          _showWarning('Material ditolak');

          await getMaterialDetail(id);

          // Refresh list if project selected
          if (selectedProject.value != null) {
            await fetchMaterialsByProject(selectedProject.value!.id);
          }
        }
      }
    } catch (e) {
      _showError(e.toString());
    } finally {
      isLoading.value = false;
    }
  }

  bool _validateForm() {
    if (selectedProject.value == null) {
      _showError('Pilih project terlebih dahulu');
      return false;
    }

    if (materialNameController.text.trim().isEmpty) {
      _showError('Nama material tidak boleh kosong');
      return false;
    }

    if (quantityController.text.isEmpty) {
      _showError('Jumlah tidak boleh kosong');
      return false;
    }

    try {
      double.parse(quantityController.text);
    } catch (e) {
      _showError('Jumlah harus berupa angka');
      return false;
    }

    if (unitController.text.trim().isEmpty) {
      _showError('Satuan tidak boleh kosong');
      return false;
    }

    return true;
  }

  void _clearForm() {
    materialNameController.clear();
    quantityController.clear();
    unitController.clear();
    descriptionController.clear();
    selectedProject.value = null;
  }

  void _showSuccess(String message) {
    Get.snackbar(
      'Berhasil',
      message,
      snackPosition: SnackPosition.TOP,
      backgroundColor: Colors.green[600],
      colorText: Colors.white,
      duration: const Duration(seconds: 3),
      margin: const EdgeInsets.all(16),
      borderRadius: 8,
      icon: const Icon(Icons.check_circle, color: Colors.white),
    );
  }

  void _showWarning(String message) {
    Get.snackbar(
      'Informasi',
      message,
      snackPosition: SnackPosition.TOP,
      backgroundColor: Colors.orange[600],
      colorText: Colors.white,
      duration: const Duration(seconds: 3),
      margin: const EdgeInsets.all(16),
      borderRadius: 8,
      icon: const Icon(Icons.info, color: Colors.white),
    );
  }

  void _showError(String message) {
    Get.snackbar(
      'Error',
      message,
      snackPosition: SnackPosition.TOP,
      backgroundColor: Colors.red[600],
      colorText: Colors.white,
      duration: const Duration(seconds: 3),
      margin: const EdgeInsets.all(16),
      borderRadius: 8,
      icon: const Icon(Icons.error, color: Colors.white),
    );
  }

  // Check permissions
  bool get canApproveReject =>
      authService.isAdmin || authService.isKepalaProyek;

  bool get canCreateRequest => authService.isMandor;

  bool get isReadOnly => authService.isPengguna;
}

File: app/Controllers/Api/MaterialController.php
<?php

namespace App\Controllers\Api;

use App\Controllers\BaseController;
use App\Models\MaterialModel;
use App\Models\ProjectModel;
use CodeIgniter\API\ResponseTrait;

class MaterialController extends BaseController
{
    use ResponseTrait;

    protected $materialModel;
    protected $projectModel;
    
    public function __construct()
    {
        $this->materialModel = new MaterialModel();
        $this->projectModel = new ProjectModel();
    }
    
    public function create()
    {
        $user = $this->request->userData ?? null;
        
        if (!$user) {
            return $this->failUnauthorized('Token invalid atau sesi berakhir');
        }

        if ($user->role !== 'mandor') {
            return $this->failForbidden('Hanya mandor yang dapat request material');
        }
        
        $rules = [
            'project_id'    => 'required|numeric',
            'material_name' => 'required|min_length[3]',
            'quantity'      => 'required|decimal',
            'unit'          => 'required'
        ];
        
        if (!$this->validate($rules)) {
            return $this->failValidationErrors($this->validator->getErrors());
        }
        
        $project = $this->projectModel->find($this->request->getVar('project_id'));
        
        if (!$project) {
            return $this->failNotFound('Project tidak ditemukan');
        }
        
        if ($project['mandor_id'] != $user->id) {
            return $this->failForbidden('Anda bukan mandor dari project ini');
        }
        
        $data = [
            'project_id'    => $this->request->getVar('project_id'),
            'mandor_id'     => $user->id,
            'material_name' => $this->request->getVar('material_name'),
            'quantity'      => $this->request->getVar('quantity'),
            'unit'          => $this->request->getVar('unit'),
            'description'   => $this->request->getVar('description'),
            'status'        => 'pending'
        ];
        
        if ($this->materialModel->insert($data)) {
            $materialId = $this->materialModel->getInsertID();
            
            $material = $this->materialModel->getMaterialWithRelations($materialId); 
            
            return $this->respondCreated([
                'status' => 'success',
                'message' => 'Request material berhasil dibuat',
                'data' => $material
            ]);
        }
        
        return $this->failServerError('Gagal membuat request material');
    }

    public function getByProject($projectId)
    {
        $user = $this->request->userData ?? null;
        if (!$user) return $this->failUnauthorized();
        
        $project = $this->projectModel->find($projectId);
        if (!$project) {
            return $this->failNotFound('Project tidak ditemukan');
        }
        
        if ($user->role !== 'admin') {
            $hasAccess = ($project['user_id'] == $user->id) ||
                         ($project['kepala_proyek_id'] == $user->id) ||
                         ($project['mandor_id'] == $user->id);
            
            if (!$hasAccess) {
                return $this->failForbidden('Akses ditolak ke project ini');
            }
        }
        
        $materials = $this->materialModel->getMaterialsWithRelations(['material_requests.project_id' => $projectId]);
        
        return $this->respond([
            'status' => 'success',
            'data' => $materials
        ]);
    }

    public function show($id)
    {
        $material = $this->materialModel->getMaterialWithRelations($id);
        
        if (!$material) {
            return $this->failNotFound('Request material tidak ditemukan');
        }
        
        return $this->respond([
            'status' => 'success',
            'data' => $material
        ]);
    }

    public function approve($id)
    {
        $user = $this->request->userData ?? null;
        if (!$user) return $this->failUnauthorized();
        
        if (!in_array($user->role, ['kepala_proyek', 'admin'])) {
            return $this->failForbidden('Akses ditolak');
        }
        
        $material = $this->materialModel->find($id);
        if (!$material) {
            return $this->failNotFound('Request material tidak ditemukan');
        }
        
        if ($user->role === 'kepala_proyek') {
            $project = $this->projectModel->find($material['project_id']);
            if ($project['kepala_proyek_id'] != $user->id) {
                return $this->failForbidden('Anda bukan Kepala Proyek di project ini');
            }
        }
        
        $data = [
            'status'      => 'approved',
            'approved_by' => $user->id,
            'approved_at' => date('Y-m-d H:i:s')
        ];
        
        if ($this->materialModel->update($id, $data)) {
            $updated = $this->materialModel->getMaterialWithRelations($id);
            return $this->respond([
                'status' => 'success',
                'message' => 'Request material berhasil disetujui',
                'data' => $updated
            ]);
        }
        
        return $this->failServerError('Gagal menyetujui request material');
    }
    
    public function reject($id)
    {
        $user = $this->request->userData ?? null;
        if (!$user) return $this->failUnauthorized();
        
        if (!in_array($user->role, ['kepala_proyek', 'admin'])) {
            return $this->failForbidden('Akses ditolak');
        }
        
        $material = $this->materialModel->find($id);
        if (!$material) {
            return $this->failNotFound('Request material tidak ditemukan');
        }
        
        $data = [
            'status'      => 'rejected',
            'approved_by' => $user->id,
            'approved_at' => date('Y-m-d H:i:s')
        ];
        
        if ($this->materialModel->update($id, $data)) {
            $updated = $this->materialModel->getMaterialWithRelations($id);
            return $this->respond([
                'status' => 'success',
                'message' => 'Request material ditolak',
                'data' => $updated
            ]);
        }
        
        return $this->failServerError('Gagal menolak request material');
    }
}